### PR TITLE
bug 1746940: document codeinfo lookup in download API

### PR DESCRIPTION
**Pull request by willkg**

### CODE REVIEW

**Comment by willkg** [comment]
@smarnach This isn't urgent. But when you have a free moment, can you review this documentation change? It's related to bug 1746940--specifically the way I overloaded the download API.

To make the documentation, you can do:

```
make docs
```

After that, I do `xdg-open docs/_build/html/download.html` to view the result in my browser. I think you said you're using Ubuntu, so that'll work for you, too (I think).

Things to review:

1. Does the documentation changes make sense?
2. Is there a better way to document the variations of this API that would be clearer for users? (try vs. non-try, debuginfo vs. codeinfo, head vs. get)

**Review by smarnach — COMMENTED** [review]
I've got a few comments, some of them not related to your changes.

I think we should also explicitly mention that HEAD requests have non-standard semantics for Tecken. [The MDN documentation of the HEAD method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) states

> The HTTP `HEAD` method requests the headers that would be returned if the `HEAD` request's URL was instead requested with the HTTP `GET` method.

This is not what Tecken is doing. While the documentation describes accurately what Tecken is doing instead, I think we should explicitly acknowledge that this is a bit unusual. We could do that in a separate PR, but it's just adding a single sentence, so we could also do it in this PR.

**Comment by willkg** [comment]
> I've got a few comments, some of them not related to your changes.
> 
> I think we should also explicitly mention that HEAD requests have non-standard semantics for Tecken. [The MDN documentation of the HEAD method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) states
> 
> > The HTTP `HEAD` method requests the headers that would be returned if the `HEAD` request's URL was instead requested with the HTTP `GET` method.
> 
> This is not what Tecken is doing. While the documentation describes accurately what Tecken is doing instead, I think we should explicitly acknowledge that this is a bit unusual. We could do that in a separate PR, but it's just adding a single sentence, so we could also do it in this PR.

I've been thinking about this. Earlier, I was of the mind that we should leave things as they are because to change the semantics would require figuring out who's relying on the current behavior, changing their code, changing the behavior, etc and in this situation where we don't have a versioned API, when we change it, we change it for everyone in a way that's not particularly discoverable and it could be a mess. But now I'm thinking we should just fix HEAD to be more correct. It still returns 200/404 based on whether the symbol exists. The change here would be to make the HTTP headers the same as the GET request and I can't think of a way someone would use HEAD such that this change would break things.

I wrote up a bug to look at fixing it. https://bugzilla.mozilla.org/show_bug.cgi?id=1861044

File: docs/download.rst [hunk 1]
```diff
@@ -78,13 +78,41 @@ Downloading API
 
 .. http:head:: /try/<DEBUG_FILENAME>/<DEBUG_ID>/<SYMBOL_FILE>
 
-   Same as ``HEAD /<DEBUG_FILENAME>/<DEBUG_ID>/<SYMBOL_FILE>``, but for try symbols.
+   Same as ``HEAD /<DEBUG_FILENAME>/<DEBUG_ID>/<SYMBOL_FILE>``, but this will
+   check try symbols and then regular symbols.
```
[the remark this answers is not in the document]
Comment (willkg) [reply docs/download.rst:82]: You're right in that there's a mix of documentation styles as well as different facts in different places and it's inconsistent. I'll do a consistency pass.

File: docs/download.rst [hunk 2]
```diff
@@ -78,13 +78,41 @@ Downloading API
...
+      < ETag: "e2e35ff973763bcda524f147981008bc"
+      < Content-Encoding: gzip
+      < Content-Type: text/plain
+      < Content-Length: 143395908
+      <
+      <OUTPUT>
+
    :reqheader Debug: if ``true``, includes a ``Debug-Time`` header in the response.
 
       If ``Debug-Time`` is `0.0``: symbol file ends in an unsupported extension
```
[the remark this answers is not in the document]
Comment (willkg) [reply docs/download.rst:118]: Oh, I see. The linter I have doesn't understand the `.. http:head::` directive, so it doesn't even look at the content block.

**Comment by willkg** [comment]
^^^ I started with your comments which were really helpful and then that turned into a minor overhaul of the entire document. This reduces a lot of redundant material, fixes a lot of consistency and clarity problems around different kinds of symbols files and how to get them, fixes path specification to what sphinxcontrib.httpdomain expects so we can do references, clarifies that hex parts need to be hex characters all upper-cased, adds example HTTP requests, adds user-agent everywhere, etc.

What do you think about these changes? Is the material easier to follow and reference?

**Review by smarnach — APPROVED** [review]
Nice, the new version makes everything really clear!

**Pull request merged**